File: jlpt/native_ios/Sources/Presentation/Study/StudyViewModel.swift
import Foundation

@MainActor
final class StudyViewModel: ObservableObject {
    @Published var level: JlptLevel = .n5 { didSet { resetAndLoad() } }
    @Published var type: StudyItemType = .vocabulary { didSet { resetAndLoad() } }
    @Published private(set) var vocabItems: [StudyItem<Vocabulary>] = []
    @Published private(set) var kanjiItems: [StudyItem<Kanji>] = []
    @Published private(set) var grammarItems: [StudyItem<Grammar>] = []
    @Published var currentIndex: Int = 0
    @Published var isFlipped = false
    @Published private(set) var isLoading = false
    @Published var error: String?

    private let getVocabularyStudy: GetVocabularyStudyUseCase
    private let getKanjiStudy: GetKanjiStudyUseCase
    private let getGrammarStudy: GetGrammarStudyUseCase
    private let updateStudyProgress: UpdateStudyProgressUseCase

    var currentCount: Int {
        switch type {
        case .vocabulary: return vocabItems.count
        case .kanji: return kanjiItems.count
        case .grammar: return grammarItems.count
        }
    }

    init(container: AppContainer) {
        getVocabularyStudy = container.getVocabularyStudy
        getKanjiStudy = container.getKanjiStudy
        getGrammarStudy = container.getGrammarStudy
        updateStudyProgress = container.updateStudyProgress
        Task { await load() }
    }

    private func resetAndLoad() {
        currentIndex = 0
        isFlipped = false
        Task { await load() }
    }

    func flip() { isFlipped.toggle() }

    func next() {
        guard currentCount > 0 else { return }
        currentIndex = (currentIndex + 1) % currentCount
        isFlipped = false
    }

    func previous() {
        guard currentCount > 0 else { return }
        currentIndex = currentIndex == 0 ? currentCount - 1 : currentIndex - 1
        isFlipped = false
    }

    func load() async {
        isLoading = true
        error = nil
        switch type {
        case .vocabulary:
            let result = await getVocabularyStudy(level: level, dueOnly: false)
            isLoading = false
            switch result {
            case .success(let items): vocabItems = items
            case .failure(let message, _): error = message
            }
        case .kanji:
            let result = await getKanjiStudy(level: level, dueOnly: false)
            isLoading = false
            switch result {
            case .success(let items): kanjiItems = items
            case .failure(let message, _): error = message
            }
        case .grammar:
            let result = await getGrammarStudy(level: level, dueOnly: false)
            isLoading = false
            switch result {
            case .success(let items): grammarItems = items
            case .failure(let message, _): error = message
            }
        }
    }

    /// Manually sets NEW/LEARNING/MASTERED for the currently shown card, independent of SRS review.
    func markStatus(_ status: StudyStatus) {
        let itemId: Int64?
        switch type {
        case .vocabulary: itemId = vocabItems[safe: currentIndex]?.item.id
        case .kanji: itemId = kanjiItems[safe: currentIndex]?.item.id
        case .grammar: itemId = grammarItems[safe: currentIndex]?.item.id
        }
        guard let itemId else { return }

        Task {
            let result = await updateStudyProgress(itemType: type, itemId: itemId, status: status)
            guard case .success = result else { return }
            let index = currentIndex
            switch type {
            case .vocabulary:
                if vocabItems.indices.contains(index) { vocabItems[index].status = status }
            case .kanji:
                if kanjiItems.indices.contains(index) { kanjiItems[index].status = status }
            case .grammar:
                if grammarItems.indices.contains(index) { grammarItems[index].status = status }
            }
            next()
        }
    }
}

extension Array {
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

File: jlpt/native_android/app/src/main/java/com/jlpt/android/presentation/study/StudyViewModel.kt
package com.jlpt.android.presentation.study

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jlpt.android.domain.model.Grammar
import com.jlpt.android.domain.model.JlptLevel
import com.jlpt.android.domain.model.Kanji
import com.jlpt.android.domain.model.StudyItem
import com.jlpt.android.domain.model.StudyItemType
import com.jlpt.android.domain.model.StudyStatus
import com.jlpt.android.domain.model.Vocabulary
import com.jlpt.android.domain.usecase.study.GetGrammarStudyUseCase
import com.jlpt.android.domain.usecase.study.GetKanjiStudyUseCase
import com.jlpt.android.domain.usecase.study.GetVocabularyStudyUseCase
import com.jlpt.android.domain.usecase.study.UpdateStudyProgressUseCase
import com.jlpt.android.domain.util.AppResult
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class StudyUiState(
    val level: JlptLevel = JlptLevel.N5,
    val type: StudyItemType = StudyItemType.VOCABULARY,
    val vocabItems: List<StudyItem<Vocabulary>> = emptyList(),
    val kanjiItems: List<StudyItem<Kanji>> = emptyList(),
    val grammarItems: List<StudyItem<Grammar>> = emptyList(),
    val currentIndex: Int = 0,
    val isFlipped: Boolean = false,
    val isLoading: Boolean = false,
    val error: String? = null
) {
    val currentCount: Int
        get() = when (type) {
            StudyItemType.VOCABULARY -> vocabItems.size
            StudyItemType.KANJI -> kanjiItems.size
            StudyItemType.GRAMMAR -> grammarItems.size
        }
}

@HiltViewModel
class StudyViewModel @Inject constructor(
    private val getVocabularyStudy: GetVocabularyStudyUseCase,
    private val getKanjiStudy: GetKanjiStudyUseCase,
    private val getGrammarStudy: GetGrammarStudyUseCase,
    private val updateStudyProgress: UpdateStudyProgressUseCase
) : ViewModel() {

    private val _uiState = MutableStateFlow(StudyUiState())
    val uiState: StateFlow<StudyUiState> = _uiState.asStateFlow()

    init {
        load()
    }

    fun selectType(type: StudyItemType) {
        _uiState.update { it.copy(type = type, currentIndex = 0, isFlipped = false) }
        load()
    }

    fun selectLevel(level: JlptLevel) {
        _uiState.update { it.copy(level = level, currentIndex = 0, isFlipped = false) }
        load()
    }

    fun flip() = _uiState.update { it.copy(isFlipped = !it.isFlipped) }

    fun next() = _uiState.update {
        if (it.currentCount == 0) it else it.copy(currentIndex = (it.currentIndex + 1) % it.currentCount, isFlipped = false)
    }

    fun previous() = _uiState.update {
        if (it.currentCount == 0) it else {
            val newIndex = if (it.currentIndex == 0) it.currentCount - 1 else it.currentIndex - 1
            it.copy(currentIndex = newIndex, isFlipped = false)
        }
    }

    fun load() {
        val level = _uiState.value.level
        val type = _uiState.value.type
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            when (type) {
                StudyItemType.VOCABULARY -> when (val result = getVocabularyStudy(level, dueOnly = false)) {
                    is AppResult.Success -> _uiState.update { it.copy(isLoading = false, vocabItems = result.data) }
                    is AppResult.Failure -> _uiState.update { it.copy(isLoading = false, error = result.message) }
                }
                StudyItemType.KANJI -> when (val result = getKanjiStudy(level, dueOnly = false)) {
                    is AppResult.Success -> _uiState.update { it.copy(isLoading = false, kanjiItems = result.data) }
                    is AppResult.Failure -> _uiState.update { it.copy(isLoading = false, error = result.message) }
                }
                StudyItemType.GRAMMAR -> when (val result = getGrammarStudy(level, dueOnly = false)) {
                    is AppResult.Success -> _uiState.update { it.copy(isLoading = false, grammarItems = result.data) }
                    is AppResult.Failure -> _uiState.update { it.copy(isLoading = false, error = result.message) }
                }
            }
        }
    }

    /** Manually sets NEW/LEARNING/MASTERED for the currently shown card, independent of SRS review. */
    fun markStatus(status: StudyStatus) {
        val state = _uiState.value
        val itemId = when (state.type) {
            StudyItemType.VOCABULARY -> state.vocabItems.getOrNull(state.currentIndex)?.item?.id
            StudyItemType.KANJI -> state.kanjiItems.getOrNull(state.currentIndex)?.item?.id
            StudyItemType.GRAMMAR -> state.grammarItems.getOrNull(state.currentIndex)?.item?.id
        } ?: return
        viewModelScope.launch {
            when (updateStudyProgress(state.type, itemId, status)) {
                is AppResult.Success -> {
                    _uiState.update {
                        when (it.type) {
                            StudyItemType.VOCABULARY -> it.copy(
                                vocabItems = it.vocabItems.mapIndexed { i, si -> if (i == it.currentIndex) si.copy(status = status) else si }
                            )
                            StudyItemType.KANJI -> it.copy(
                                kanjiItems = it.kanjiItems.mapIndexed { i, si -> if (i == it.currentIndex) si.copy(status = status) else si }
                            )
                            StudyItemType.GRAMMAR -> it.copy(
                                grammarItems = it.grammarItems.mapIndexed { i, si -> if (i == it.currentIndex) si.copy(status = status) else si }
                            )
                        }
                    }
                    next()
                }
                is AppResult.Failure -> Unit
            }
        }
    }
}
